src/lib/modules/ping/ping.go

```diff
@@ -78,26 +78,35 @@ func (ps *Ping) setConfigurations() {
 }
 
 func (ps *Ping) perform(pingInterval TestInterval) {
+	defer log.Infoln("terminating ping goroutine")
+	var wg sync.WaitGroup
+	wg.Add(len(*ps.targets))
+	for machineHash, machineIP := range *ps.targets {
+		go ps.pingExec(machineIP.IPDomain, machineHash, 3, &wg, pingInterval)
+	}
+	wg.Wait()
+}
+
+func (ps *Ping) pingExec(urlRaw, machineHash string, packets int, wg *sync.WaitGroup, pingInterval TestInterval) {
 	for {
 		switch ps.localConfig.Config.UtilsConf.ServicesSignal.Ping {
 		case "active":
 			err, _ := utils.VerifyConnection()
 			if !err {
 				log.Warnln("unable to connect external network. please check you internet connection", "p")
 			} else {
-				var wg sync.WaitGroup
-				wg.Add(len(*ps.targets))
-				for machineHash, machineIP := range *ps.targets {
-					go ps.ping(machineIP.IPDomain, machineHash, 3, &wg)
-				}
-				wg.Wait()
+				var waitgroup sync.WaitGroup
+				waitgroup.Add(1)
+				go ps.ping(urlRaw, machineHash, packets, &waitgroup)
+				waitgroup.Wait()
 			}
 		case "passive":
 			// terminate the goroutine
-			log.Infoln("terminating ping goroutine")
+			wg.Done()
 			return
 		default:
 			log.Warnln("invalid service-state value of ping")
+			wg.Done()
```

Comment:
Ideally, we should use `defer wg.Done()` as the first line of this function instead of calling `wg.Done()` at each return statement.

Right now, as `wg.Done()` is not called before line 123, there is a possibility that the parent function will never stop executing.

Reply:
Thanks for the review, @chetanyakan, but actually that is what the expected behavior should be. Also, that is the reason I didn't put wg.Done() in defer statement.
In the case of active, the parent function is not supposed to stop, instead, sleep for the specified duration and execute again till the status is not turned `passive`.

Reply:
@Tushar3099 Don't you think there should be an equal number of `wg.Add()` and `wg.Done()` calls or otherwise the program won't stop even if there's nothing to execute?

P.S. In case of active, it will keep working in the same manner as of right now, even when you call `defer wg.Done()` before the for loop.

Reply:
@chetanyakan Please be little specific here : 

> @Tushar3099 Don't you think there should be an equal number of wg.Add() and wg.Done() calls or otherwise the program won't stop even if there's nothing to execute?

I don't get what you want to say : 

> P.S. In case of active, it will keep working in the same manner as of right now, even when you call defer wg.Done() before the for loop.

- We are adding wait groups equal to the number of targets and then running the pingExec functions.
- We want them to run infinitely with sleep durations until the status is passive.
- In the case of active we do not want the function to stop, also it is obvious if we put defer statement outside the loop it will never be executed in case of active as it will never come out of the loop.
- We can do one thing to run a defer statement outside the loop and the return from the loop in case of default or passive, but that would also work the same as of right now.

If there is a possible way where we would want the function to stop but it won't stop please point it out. But in the case of active, we don't want the function to stop.

Reply:
> We can do one thing to run a defer statement outside the loop and the return from the loop in case of default or passive, but that would also work the same as of right now.

I think it would not work the same as of right now. But please correct me if I am wrong.
Please look at the code below (lines 113 - 124)
```go
		intrv := time.Duration(*ps.localConfig.Config.Interval[0].Duration)
		switch pingInterval.OfType {
		case "hr":
			time.Sleep(intrv * time.Hour)
		case "min":
			time.Sleep(intrv * time.Minute)
		case "sec":
			time.Sleep(intrv * time.Second)
		default:
			log.Infoln("invalid interval-type for ping")
			return
		}
```
What would happen if the program returns in the `default` case of `pingInterval.OfType`?

1. Would `wg.Done()` be called or not?
2. Should `wg.Done()` be called here to make sure `wg.Wait()` on line 87 can finish executing?

Reply:
That's absolutely right. Thanks for pointing out, we would require one more wg.Done(), before this default statement. However, I  think I should write wg.Done() outside loop with defer so that we don't have to put wg.Done() everywhere.